crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,516 @@
+
+                    if new_file.is_none() {
+                        let file_path = old_selectable_path.join(file_name);
+                        operations.push(FileSystemOperation::DeleteFile(file_path));
+                    }
+                }
+            }
+        }
+
+        for file_name in old.root_files.keys() {
+            if !new.root_files.contains_key(file_name) {
+                let file_path = artifact_directory.join(file_name);
+                operations.push(FileSystemOperation::DeleteFile(file_path));
+            }
+        }
+
+        operations
+    }
+}
+
+impl From<&[ArtifactPathAndContent]> for FileSystemState {
+    fn from(artifacts: &[ArtifactPathAndContent]) -> Self {
+        let mut root_files = HashMap::new();
+        let mut nested_files: HashMap<
+            ServerObjectEntityName,
+            HashMap<SelectableName, HashMap<ArtifactFileName, (Index<FileContent>, ArtifactHash)>>,
+        > = HashMap::new();
+
+        for (index, artifact) in artifacts.iter().enumerate() {
+            let value = (
+                Index::new(index),
+                ArtifactHash::from(hash(
+                    &artifact.file_content,
+                    PersistedDocumentsHashAlgorithm::Md5,
+                )),
+            );
+
+            match &artifact.artifact_path.type_and_field {
+                Some(type_and_field) => {
+                    nested_files
+                        .entry(type_and_field.parent_object_entity_name)
+                        .or_default()
+                        .entry(type_and_field.selectable_name)
+                        .or_default()
+                        .insert(artifact.artifact_path.file_name, value);
+                }
+                None => {
+                    root_files.insert(artifact.artifact_path.file_name, value);
+                }
+            }
+        }
+
+        FileSystemState {
+            root_files,
+            nested_files,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use common_lang_types::{
+        ArtifactPath, ParentObjectEntityNameAndSelectableName, SelectableName,
+        ServerObjectEntityName,
+    };
+    use intern::string_key::Intern;
+    use std::path::PathBuf;
+
+    fn create_artifact(
+        server: Option<&str>,
+        selectable: Option<&str>,
+        file_name: &str,
+        content: &str,
+    ) -> ArtifactPathAndContent {
+        let type_and_field = match (server, selectable) {
+            (Some(s), Some(sel)) => Some(ParentObjectEntityNameAndSelectableName {
+                parent_object_entity_name: ServerObjectEntityName::from(s.intern()),
+                selectable_name: SelectableName::from(sel.intern()),
+            }),
+            _ => None,
+        };
+
+        ArtifactPathAndContent {
+            artifact_path: ArtifactPath {
+                type_and_field,
+                file_name: ArtifactFileName::from(file_name.intern()),
```

Comment:
I'm not sure why the compiler was complaining, but FWIW it's not complaining when I do `sel.intern().into()`